benchmarks/tensors/dibaryon/baryon_generator.cpp

```diff
@@ -47,168 +210,329 @@ void generate_function(std::string name)
-
-    computation Q_r_init("Q_r_init", {t, n, iCprime, iSprime, kCprime, kSprime, jCprime, jSprime, x, y}, expr((double) 0));
-    computation Q_i_init("Q_i_init", {t, n, iCprime, iSprime, kCprime, kSprime, jCprime, jSprime, x, y}, expr((double) 0));
-
-    computation Bsingle_r_init("Bsingle_r_init", {t, n, iCprime, iSprime, jCprime, jSprime, kCprime, kSprime, x, x2}, expr((double) 0));
-    computation Bsingle_i_init("Bsingle_i_init", {t, n, iCprime, iSprime, jCprime, jSprime, kCprime, kSprime, x, x2}, expr((double) 0));
-
-    computation Q_r_update("Q_r_update", {t, n, iCprime, iSprime, kCprime, kSprime, jCprime, jSprime, x, y, wnum},
-			Q_r_init(t, n, iCprime, iSprime, kCprime, kSprime, color_weights(wnum, 1), spin_weights(wnum, 1), x, y) + weights(wnum) * mul_r(psi, m1));
-    Q_r_update.add_predicate((jCprime == color_weights(wnum, 1)) && (jSprime == spin_weights(wnum, 1)));
-
-    computation Q_i_update("Q_i_update", {t, n, iCprime, iSprime, kCprime, kSprime, jCprime, jSprime, x, y, wnum},
-			Q_i_init(t, n, iCprime, iSprime, kCprime, kSprime, color_weights(wnum, 1), spin_weights(wnum, 1), x, y) + weights(wnum) * mul_i(psi, m1));
-    Q_i_update.add_predicate((jCprime == color_weights(wnum, 1)) && (jSprime == spin_weights(wnum, 1)));
-
-    std::pair<expr, expr> Q_update(Q_r_update(t, n, iCprime, iSprime, kCprime, kSprime, lCprime, lSprime, x, y, wnum), Q_i_update(t, n, iCprime, iSprime, kCprime, kSprime, lCprime, lSprime, x, y, wnum));
-    std::pair<expr, expr> prop_1p(prop_r(t, 1, jCprime, jSprime, lCprime, lSprime, x2, y), prop_i(t, 1, jCprime, jSprime, lCprime, lSprime, x2, y));
-
-    computation Bsingle_r_update("Bsingle_r_update", {t, n, iCprime, iSprime, jCprime, jSprime, kCprime, kSprime, lCprime, lSprime, x, x2, y},
-	    Bsingle_r_init(t, n, iCprime, iSprime, jCprime, jSprime, kCprime, kSprime, x, x2) + mul_r(Q_update, prop_1p));
+    std::map<std::string, expr> prop_loads;
+    std::map<std::pair<int,int>, std::pair<expr, expr>> Q_exprs;
+    std::map<std::pair<int,int>, std::pair<expr, expr>> O_exprs;
+    std::map<std::pair<int,int>, std::pair<expr, expr>> P_exprs;
+    // FIRST: build the ``unrolled'' expressions of Q, O, and P
+    for (int ii = 0; ii < Nw; ii++) {
+      int ic = test_color_weights[ii][0];
+      int is = test_spin_weights[ii][0];
+      int jc = test_color_weights[ii][1];
+      int js = test_spin_weights[ii][1];
+      int kc = test_color_weights[ii][2];
+      int ks = test_spin_weights[ii][2];
+      double w = test_weights[ii];
+
+      complex_expr prop_0 = prop(t, 0, iCprime, iSprime, ic, is, x, y);
+      complex_expr prop_2 = prop(t, 2, kCprime, kSprime, kc, ks, x, y);
+      complex_expr prop_0p = prop(t, 0, kCprime, kSprime, ic, is, x, y);
+      complex_expr prop_2p = prop(t, 2, iCprime, iSprime, kc, ks, x, y);
+      
+      complex_expr q = (prop_0 * prop_2 - prop_0p * prop_2p) * w;
+      insert_or_add(Q_exprs, {jc, js}, q);
```

Comment:
Do you want to simplify this in another commit? The alternative to this would be to, as you suggested, create Q_exprs[jc][js], loop over it, and check for non-zero computations.

Reply:
Yes sure, you can do simplify it in a future commit. That works for me.